lib/ballot.go

```diff
@@ -370,6 +370,22 @@ func (b *BallotBoxes) RemoveVotingResult(vr *VotingResult) (err error) {
 	return
 }
 
+func (b *BallotBoxes) RemoveVotingResultByHash(hash string) (err error) {
+	if !b.HasMessageByHash(hash) {
+		err = sebakerror.ErrorVotingResultNotFound
+		return
+	}
+
+	b.Lock()
+	defer b.Unlock()
+
+	delete(b.Results, hash)
+	delete(b.Messages, hash)
+	// need to delete b.Sources
```

Comment:
@Charleslee522 @anarcher I want to create a method that uses `MessageHash` to clear the corresponding `VotingResult`. But it seems that the `Source` for `MessageHash` should also be cleared, and I am wondering how to fix it.

Reply:
@ enessech
Oh... b.Sources couldn’t be removed by only hash, it needs voting result.
I figure, lazy removing is better.

Reply:
I am not sure that works. When we remove VoltingResult, Can we remove `Source`? 

```
if vr,ok := b.Results[hash] ; ok {
   delete(b.Sources,vr.Source()]
}
```

Reply:
@anarcher 
```
 func (b *BallotBoxes) RemoveVotingResult(vr *VotingResult) (err error) {
 	if !b.HasMessageByHash(vr.MessageHash) {
 		err = sebakerror.ErrorVotingResultNotFound
 		return
 	}
 
 	b.Lock()
 	defer b.Unlock()
 
 	delete(b.Results, vr.MessageHash)
 	delete(b.Messages, vr.MessageHash)
 	delete(b.Sources, vr.Source)
 
  	return
  }
```
Using the method(`RemoveVotingResult`) in 357th line, we can remove `VotingResult` and `Source`

Reply:
@anarcher I will try to use this method.

Reply:
> When we remove VoltingResult, Can we remove Source?

@anarcher Yes right.

> Using the method(RemoveVotingResult) in 357th line, we can remove VotingResult and Source

@enessech I remember that we save only Message hash and VotingResult hash, but now we need to save VotingResult.Source() also.